Paginating VAL app views front end and backend

shahbazn opened the pull request:
- Adding Backend paginations support at db level
- Client side pagination of video upload views similar
   Files & Upload views with sorting and filtering support
 - Only previous video upload views will be paginated 
 - Active video upload view will remain same and decoupled
    from previous video upload view

- [ ] Build a sandbox for your branch and add a link here

- [ ] i18n
- [ ] RTL
- [ ] safecommit violation code review process
- [ ] Unit, integration, acceptance tests as appropriate
- [ ] Analytics
- [ ] Performance
- [ ] Database migrations are backwards-compatible

If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review: @mushtaqak
- [ ] Code review: @muhammad-ammar
- [ ] Doc Review: @...
- [ ] UX review: @...
- [ ] Accessibility review: @...
- [ ] Product review: @...

If it's needed, or even if you suspect it might be needed, do any/all these things as early as possible. Don't surprise devops right before merging!
For example, touching envs/aws.py or or envs/common.py are dead giveaways that you'll need assistance when pushing your changes out.
Things that require devops attention include, but are not limited to:
- configuration changes
- feature flags
- migrations
Ways to get devops attention, pick-and-choose whichever is most appropriate for your specific PR:
- [ ] Mention your concerns to Max in standup, or any other member of devops face-to-face when convinient
- [ ] Tag devops on your PR (@ edx/devops, or @ maxrothman), either for review (more specific to devops = better) or just a general FYI
- [ ] File a support ticket by emailing support@edx.org

FYI: Tag anyone who might be interested in this PR here.

- [ ] Rebase and squash commits

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

Commit:
This is a work in progress will be updates
 - front end views wrapped in pagination views from ux-toolkit
 - header styling is broken needs to be fixed
 - footer header navigation is broken and needs to be fixed

- [ ] Build a sandbox for your branch and add a link here

- [ ] i18n
- [ ] RTL
- [ ] safecommit violation code review process
- [ ] Unit, integration, acceptance tests as appropriate
- [ ] Analytics
- [ ] Performance
- [ ] Database migrations are backwards-compatible

If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review: @Muzaffar
- [ ] Code review: @Ammar
- [ ] Doc Review: @...
- [ ] UX review: @...
- [ ] Accessibility review: @...
- [ ] Product review: @...

If it's needed, or even if you suspect it might be needed, do any/all these things as early as possible. Don't surprise devops right before merging!
For example, touching envs/aws.py or or envs/common.py are dead giveaways that you'll need assistance when pushing your changes out.
Things that require devops attention include, but are not limited to:
- configuration changes
- feature flags
- migrations
Ways to get devops attention, pick-and-choose whichever is most appropriate for your specific PR:
- [ ] Mention your concerns to Max in standup, or any other member of devops face-to-face when convinient
- [ ] Tag devops on your PR (@ edx/devops, or @ maxrothman), either for review (more specific to devops = better) or just a general FYI
- [ ] File a support ticket by emailing support@edx.org

FYI: Tag anyone who might be interested in this PR here.

- [ ] Rebase and squash commits

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

Commit:
WIP [TNL-5941]
 - common/js/components/views/paging_header did not
   have navigation controls template so changes paging
   header wrapper to
   js/views/paging_header

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

Commit:
[TNL-5941] Adding pagination to VAL views

 - Added pagination to video upload views
 - addes pagination support for get videos APIs in platform

Signed-off-by: Shahbaz Nazir <shahbaznazir@ShahbazNazir-MBP-113.arbisoft.com>

shahbazn committed:
Update videos_index.js

Commit:
[TNL-5941] pagination for VAL app

Signed-off-by: Shahbaz Nazir <shahbaznazir@ShahbazNazir-MBP-113.arbisoft.com>

Commit:
Merge branch 'TNL-5941' of https://github.com/shahbazn/edx-platform into TNL-5941

Commit:
[TNL-5941] fix collection binding

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

muzaffaryousaf commented on cms/djangoapps/contentstore/views/videos.py:
```diff
@@ -10,6 +10,7 @@
 
 from django.conf import settings
 from django.contrib.auth.decorators import login_required
+from django.core.paginator import Paginator
```
Do we still need this import ? I think we're not using it anymore.

shahbazn replied:
no will remove this

muzaffaryousaf commented on cms/djangoapps/contentstore/views/videos.py:
```diff
@@ -285,33 +286,41 @@ def convert_video_status(video):
-    videos = list(get_videos_for_course(course.id, VideoSortField.created, SortDirection.desc))
+    params = request.GET.dict()
 
+    resp = get_videos_for_course(course.id, **params)
```
I would rename it to `videos_data` since it also includes  paging information.

muzaffaryousaf commented on cms/djangoapps/contentstore/views/videos.py:
```diff
@@ -285,33 +286,41 @@ def convert_video_status(video):
     Retrieves the list of videos from VAL corresponding to this course.
     """
-    videos = list(get_videos_for_course(course.id, VideoSortField.created, SortDirection.desc))
+    params = request.GET.dict()
```
I would prefer to send explicit parameters to avoid any unwanted param from front-end.

muzaffaryousaf commented on cms/static/js/collections/video.js:
```diff
@@ -0,0 +1,67 @@
+        initialize: function(models, options){
+            PagingCollection.prototype.initialize.call(this, models, options);
+            this.url = options.url;
+        },
```
Why we need both methods `constructor` & `initialize` ?

muzaffaryousaf commented on cms/static/js/collections/video.js:
```diff
@@ -0,0 +1,67 @@
+        },
+
+        parse: function(response, options) {
+            response.results = response.videos;
```
Ideally, the list of videos should be in `results`.

muzaffaryousaf replied:
Backend should send list in results for consistency across platform

muzaffaryousaf commented on cms/static/js/factories/videos_index.js:
```diff
@@ -26,7 +27,15 @@ define([
+                            page_size: responseData.page_size,
+                            sort_field: responseData.sort_field,
+                            total_count: responseData.total_count,
+                            sort_dir: responseData.sort_dir
```
We should use camel case naming convention for js variables / methods.

muzaffaryousaf commented on cms/static/js/factories/videos_index.js:
```diff
@@ -44,7 +53,15 @@ define([
+                    page_size: previousUploads["page_size"],
+                    sort_field: previousUploads["sort_field"],
+                    total_count: previousUploads["total_count"],
+                    sort_dir: previousUploads["sort_dir"]  
```
Again, use camel case for keys of dict/json.

muzaffaryousaf reviewed (changes requested):
@shahbazn i've gone through code roughly and left some feedback, will continue my review after the sandbox, you need to set edx-val dependency in requirements first. Once you're done let @mushtaqe know he'll build a sandbox.

Commit:
Improve video paging collection

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

Commit:
Minor bug fixes

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

Commit:
updating unit tests

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

Commit:
updating video unit test

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

Commit:
adding js unit tests for val pagination

 - also includes minor bug fix

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

Commit:
updating js unit test

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

muzaffaryousaf commented on cms/djangoapps/contentstore/views/tests/test_videos.py:
```diff
@@ -177,11 +177,11 @@ class VideosHandlerTestCase(VideoUploadTestMixin, CourseTestCase):
+        response_videos = response_videos_data["results"]
         self.assertEqual(len(response_videos), len(self.previous_uploads))
         for i, response_video in enumerate(response_videos):
-            # Videos should be returned by creation date descending
```
Are we not assuming this case anymore ?

shahbazn replied:
no

muzaffaryousaf commented on cms/djangoapps/contentstore/views/tests/test_videos.py:
```diff
@@ -556,8 +558,7 @@ def _check_csv_response(self, expected_profiles):
             }
-            # Videos should be returned by creation date descending
-            original_video = self.previous_uploads[-(i + 1)]
+            original_video = self.previous_uploads[i]
```
Where are the tests related to changes that you've made in videos.py ?

shahbazn replied:
no new functionality added. updated the test case to test for new parameters.

muzaffaryousaf replied:
can we explicitly check the ordering here.

muzaffaryousaf commented on cms/djangoapps/contentstore/views/videos.py:
```diff
@@ -285,33 +286,44 @@ def convert_video_status(video):
-    videos = list(get_videos_for_course(course.id, VideoSortField.created, SortDirection.desc))
+    params = {
+        "page": request.GET.get("page", 1),
+        "page_size": request.GET.get("page_size", 20),
```
The default size should come from some settings.

muzaffaryousaf replied:
On 2nd thought we should not specify default page size here. It should be defined in Val.

muzaffaryousaf commented on cms/static/js/spec/views/previous_video_upload_list_spec.js:
```diff
@@ -99,6 +107,121 @@ define(
                 verifyVideosInfo(this, false);
             });
+
+            describe('Paging Previous Video Upload List View', function() {
```
Generally we don't use spaces in describe naming.

shahbazn replied:
done

muzaffaryousaf commented on cms/static/js/spec/views/previous_video_upload_list_spec.js:
```diff
@@ -99,6 +107,121 @@ define(
+
+            describe('Paging Previous Video Upload List View', function() {
+
+                var videoView
```
Don't forget to add semi-colon for js statements.

shahbazn replied:
done

muzaffaryousaf commented on cms/static/js/spec/views/previous_video_upload_list_spec.js:
```diff
@@ -99,6 +107,121 @@ define(
+                var videoView
+
+                beforeEach(function() {
+                    var collection = new VideoPagingCollection([
```
I would rename collection to something more meaningful. VideosCollection maybe .

muzaffaryousaf replied:
I would declare this variable in start of describe along with videoView.

shahbazn replied:
done

muzaffaryousaf commented on cms/static/js/spec/views/previous_video_upload_list_spec.js:
```diff
@@ -99,6 +107,121 @@ define(
+                        {
+                            url: videoHandlerUrl,
+                            pageSize: 2,
+                            sortField: "Name",
```
Let's be consistent with usage of quote, single / double.

shahbazn replied:
done

muzaffaryousaf commented on cms/djangoapps/contentstore/views/videos.py:
```diff
@@ -285,33 +286,44 @@ def convert_video_status(video):
     """
-    videos = list(get_videos_for_course(course.id, VideoSortField.created, SortDirection.desc))
+    params = {
+        "page": request.GET.get("page", 1),
```
better to use single quote in python.

shahbazn replied:
its double quotes through out keeping it consistent

muzaffaryousaf commented:
Link to sandbox https://studio-video-paging.sandbox.edx.org/videos/course-v1:edX+DemoX+Demo_Course

muzaffaryousaf commented:
@shahbazn you need to rebase these PRs.

Commit:
Minor code improvements

 - sorting for created column

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

Commit:
Updating version for edx-val

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

muzaffaryousaf commented on requirements/edx/github.txt:
```diff
@@ -79,7 +79,7 @@ git+https://github.com/edx/XBlock.git@xblock-0.4.13#egg=XBlock==0.4.13
 -e git+https://github.com/edx/edx-submissions.git@1.1.4#egg=edx-submissions==1.1.4
 git+https://github.com/edx/ease.git@release-2015-07-14#egg=ease==0.1.3
-git+https://github.com/edx/edx-val.git@0.0.12#egg=edxval==0.0.12
+git+https://github.com/shahbazn/edx-val.git@0.0.13#egg=edxval==0.0.13
```
Specify your branch's commit SHA here.

muzaffaryousaf commented on cms/static/js/collections/video.js:
```diff
@@ -0,0 +1,55 @@
+
+    var VideoPagingCollection = PagingCollection.extend({
+
+        initialize: function(models, options){
```
Leave empty space like `options) {`

shahbazn replied:
done

Commit:
adding latest commit head

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

Commit:
Minor fixes

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

shahbazn commented:
rebased

muzaffaryousaf commented:
http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/javascript_guidelines.html
@shahbazn FYI ^^

muzaffaryousaf commented:
On paging footer it says `Showing 0-0 out of 0 total, sorted by Date Added ascending`. I think it should be in descending order by default to see latest videos first.

shahbazn commented:
@muzaffaryousaf agreed will update default sorting order

muzaffaryousaf commented:
@shahbazn I'm seeing `#` in url when i click on sort field, can you look into that. it should't show the `#` sign in url.

muzaffaryousaf commented:
@shahbazn Can we also add sorting to Name,Duration and Status fields ?

muzaffaryousaf commented:
The look and behavior should be just like existing paginated features e.g. Files & Uploads and Notes.

muzaffaryousaf commented:
@shahbazn Have you handled "Download available encodings (.csv)" feature on video Uploads page ?

muzaffaryousaf commented on cms/static/js/views/previous_video_upload_list.js:
```diff
@@ -1,30 +1,76 @@
+                    this.template = options.template;
+                },
+
+                renderPageItems: function(){
```
Leave a space between` () & {` and change all occurrences.

shahbazn replied:
done

muzaffaryousaf commented on cms/static/js/views/previous_video_upload_list.js:
```diff
@@ -1,30 +1,76 @@
+                    $el.html(this.template({encodingsDownloadUrl: this.encodingsDownloadUrl}));
+                    $tabBody = $el.find('.js-table-body');
+                    _.each(this.itemViews, function(view) {
+                                   $tabBody.append(view.render().$el);
```
Correct the indentation.

shahbazn replied:
done

muzaffaryousaf commented on cms/static/js/views/previous_video_upload_list.js:
```diff
@@ -1,30 +1,76 @@
+                                   $tabBody.append(view.render().$el);
+                    });
+                    this.pagingHeader = new PagingHeader({view: this,
+                        el: $el.find('#video-paging-header')});
```
Can't we specify this on 1 line ?

shahbazn replied:
done

muzaffaryousaf commented on cms/static/js/views/previous_video_upload_list.js:
```diff
@@ -1,30 +1,76 @@
+                    this.pagingHeader = new PagingHeader({view: this,
+                        el: $el.find('#video-paging-header')});
+                    this.pagingFooter = new PagingFooter({collection: this.collection,
+                        el: $el.find('#video-paging-footer')});
```
Why not on 1 line.

shahbazn replied:
done

muzaffaryousaf commented on cms/templates/js/previous-video-upload-list.underscore:
```diff
@@ -5,17 +5,24 @@
         <thead>
             <tr>
                 <th><%- gettext("Name") %></th>
                 <th><%- gettext("Duration") %></th>
```
I'd add sorting to Name, duration and status.

muzaffaryousaf commented:
@shahbazn Open this PR against actual platform master branch.

muzaffaryousaf commented on cms/djangoapps/contentstore/views/videos.py:
```diff
@@ -285,33 +286,46 @@ def convert_video_status(video):
     """
-    videos = list(get_videos_for_course(course.id, VideoSortField.created, SortDirection.desc))
+    params = {
+        "page": request.GET.get("page", 1),
```
The default values should be in edx val.

muzaffaryousaf commented on cms/djangoapps/contentstore/views/videos.py:
```diff
@@ -171,7 +171,8 @@ def get_profile_header(profile):
 
-    videos = list(_get_videos(course))
+    videos_data = _get_videos(course, request)
+    videos = videos_data["results"]
```
I would leave the download functionality as it for now. i.e. don't get the paginated response instead of just returning first page only. Is it possible ?

shahbazn replied:
done

muzaffaryousaf commented on cms/djangoapps/contentstore/views/videos.py:
```diff
@@ -285,33 +286,46 @@ def convert_video_status(video):
+    videos_data = get_videos_for_course(course.id, **params)
+    videos = videos_data["results"]
     # convert VAL's status to studio's Video Upload feature status.
     for video in videos:
```
I would loop on videos_data["results"] directly instead of assigning it to `videos` and then re-assigning it.

muzaffaryousaf commented on cms/djangoapps/contentstore/views/videos.py:
```diff
@@ -285,33 +286,46 @@ def convert_video_status(video):
+        for video in videos_data["results"]
     )
 
+    videos_data["results"] = videos
```
You don't need to re-assign it since you're changing the videos_data["results"] directly.

shahbazn replied:
we are creating a new list with specific fields and discarding any other fields.

muzaffaryousaf commented on cms/djangoapps/contentstore/views/videos.py:
```diff
@@ -134,9 +134,9 @@ def videos_handler(request, course_key_string, edx_video_id=None):
+            return videos_index_json(course, request)
         else:
-            return videos_index_html(course)
+            return videos_index_html(course, request)
```
I'd choose request as first parameter.

shahbazn replied:
keeping it consistent with existing code e.g. videos_post method uses course, request

muzaffaryousaf commented on cms/djangoapps/contentstore/views/tests/test_videos.py:
```diff
@@ -177,11 +177,17 @@ class VideosHandlerTestCase(VideoUploadTestMixin, CourseTestCase):
     def test_get_json(self):
```
You might need to add doc string here to pass the quality check.

muzaffaryousaf commented on cms/djangoapps/contentstore/views/tests/test_videos.py:
```diff
@@ -177,11 +177,17 @@ class VideosHandlerTestCase(VideoUploadTestMixin, CourseTestCase):
         for i, response_video in enumerate(response_videos):
-            # Videos should be returned by creation date descending
-            original_video = self.previous_uploads[-(i + 1)]
+            original_video = self.previous_uploads[i]
```
@shahbazn Can we verify that the videos are in ordered by "created" in asc order. I know we've checked the value in response data but that's not sufficient.

shahbazn replied:
that coverage already added in val can be added here as well though

muzaffaryousaf commented:
@shahbazn When i delete video it should update the results count in header.

muzaffaryousaf commented on cms/djangoapps/contentstore/views/videos.py:
```diff
@@ -321,18 +335,23 @@ def videos_index_html(course):
+        "end": <last_page>,
+        "total_count": <total number of records>,
+        "sort_field": <sort field>,
+        "sort_dir": <sort direction>,
```
I can't see start and end in json response.

muzaffaryousaf commented on cms/static/js/collections/video.js:
```diff
@@ -0,0 +1,55 @@
+    var VideoPagingCollection = PagingCollection.extend({
+
+        initialize: function(models, options) {
+            this.state = this.initState(options);
```
Can't we just set the state outside the initialize method, instead of `initState` method.

shahbazn replied:
we need this to set the initial state values from options.

muzaffaryousaf commented on cms/static/js/collections/video.js:
```diff
@@ -0,0 +1,55 @@
+                    } else {
+                        return -1
+                    }
+                }
```
```
if ( options.sort_dir === 'desc') {
    return 1
} else {
    return -1
}
```

muzaffaryousaf replied:
btw why do we need this explicit order method ?

shahbazn replied:
removed dont need it

muzaffaryousaf commented on cms/static/js/collections/video.js:
```diff
@@ -0,0 +1,55 @@
+define([
```
I'd add doc string for this file which tells that this video collection is for previous uploaded videos since we're using general file name.

muzaffaryousaf commented on cms/static/js/collections/video.js:
```diff
@@ -0,0 +1,55 @@
+            currentPage: 'page',
+            pageSize: 'page_size',
+            sortKey: 'sort_field',
+            order: 'sort_dir',
```
`sort_dir` --> `direction` ?

shahbazn replied:
yes

muzaffaryousaf replied:
can we not rename this ?

muzaffaryousaf commented on cms/static/js/factories/videos_index.js:
```diff
@@ -1,7 +1,8 @@
 define([
-    'jquery', 'backbone', 'js/views/active_video_upload_list',
+    'jquery', 'backbone', 'js/views/active_video_upload_list', 'js/collections/video',
```
I would add this to the end of the list after all views.

shahbazn replied:
done

muzaffaryousaf commented on cms/static/js/factories/videos_index.js:
```diff
@@ -26,7 +27,13 @@ define([
                         type: 'GET'
                     }).done(function(responseData) {
-                        var updatedCollection = new Backbone.Collection(responseData.videos).filter(function(video) {
+                        var updatedCollection = new VideoPagingCollection(responseData.videos, {
```
Do we still need to use `responseData.videos` ?

shahbazn replied:
no we don't. fixed

Commit:
adding my self to authers

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

Commit:
Adding sorting for name and duration fields

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

shahbazn commented:
@muzaffaryousaf added sorting to Name,Duration. Status not supported.

muzaffaryousaf commented:
@shahbazn Let me know when you're done with all feedback comments.

Commit:
fix encoded video downloads csv file

 - currently it always downloads the first page
 - should dowload entire data unpaginated

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

Commit:
Fix download available encodings

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

shahbazn commented:
@muzaffaryousaf 
>@shahbazn Have you handled "Download available encodings (.csv)" feature on video Uploads page ?

done

shahbazn commented:
>On paging footer it says Showing 0-0 out of 0 total, sorted by Date Added ascending. I think it should be in descending order by default to see latest videos first.

done

Commit:
Bug fixes Addressing review comments

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

Commit:
updating tests

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

Commit:
minor fixes

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

Commit:
updating edv-val dependancy commit hash

Signed-off-by: Shahbaz Nazir <shahbaz.nazir@arbisoft.com>

muzaffaryousaf commented:
In favor of https://github.com/edx/edx-platform/pull/14605
